### PR TITLE
The cash box finder component

### PR DESCRIPTION
**Pull request by DedrickEnc**
The cash box finder component is let the user choose the cash and send back the account id of the chosen cash.

that how to use it in the page : 

``` html
<div class="form-group">
  <label class="control-label">{{ "YOUR_KEY" | translate }}</label>
  <bh-cash-box-finder cash-box-value="yourModel"></bh-cash-box-finder>
</div>
```

### CODE REVIEW

**Comment by DedrickEnc** [comment]
@jniles

File: .gitignore [hunk 1]
```diff
@@ -29,3 +29,4 @@ client/src/upload/
 # ignore configuration
 .env.production
 .env.development
+integrations.log
```
Comment (jniles) [inline .gitignore:32]: This was probably generated by your LOG_FILE variable in `.env.development` or `.env.production`.  Check to see if that is the case.

File: client/src/js/components/bhCashBoxFinder.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+ * @description a component to deal with cash box, it let you choose a cash box
+ * 
+ **/
+ 
+ //A cash box finder component controller
+ 
+ function cashBoxFinderController (CashboxService, $translate){
+ 	
+ 	var ctrl = this;
+ 	ctrl.types = [
```
Comment (jniles) [inline client/src/js/components/bhCashBoxFinder.js:15]: In this case, it is probably better to simply have the label as the untranslated key and then translate it in the HTML.  That way, your test cases will be more predictable.

File: client/src/js/components/bhCashBoxFinder.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+ 				 ];
+ 	ctrl.selectedType = ctrl.types[0];
+
+ 	function init () {
+ 		ctrl.selectedCash = null;
+ 		ctrl.cashBoxValue = null;
+	 	ctrl.session = {
+	 		state : 'finding'
+	 	}
+	 	fetchcash('?full=1&is_auxillary=' + ctrl.selectedType.is_auxillary + '&is_bank=' + ctrl.selectedType.is_bank);
```
Comment (jniles) [inline client/src/js/components/bhCashBoxFinder.js]: Instead of manually writing query strings, you should use Angular's `$http` parameter serializer.  [Here is the documentation](https://docs.angularjs.org/api/ng/service/$http).

Essentially, you should be able to do this:

``` js
var opts = {
  is_auxillary : 0
};

$http.get('/cashboxes', { params : opts }); // this will send GET /cashboxes?is_auxillary=0
```

File: client/src/partials/templates/bhCashBoxFinder.tmpl.html [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+      </ul>
+    </div>
+
+    <input 
+      bh-cash-box-finder      
+      name="cashBoxFinderValue"
+      ng-model="$ctrl.value"    
+      type="text"
+      class="form-control"
+      uib-typeahead="cashBox as cashBox.text for cashBox in $ctrl.search($viewValue)"
```
Comment (jniles) [inline client/src/partials/templates/bhCashBoxFinder.tmpl.html:23]: Why don't you make the `search()` function send an `$http` request and do the filtering in the database?  That way, you do not need to use a regex on the client side.

File: server/controllers/medical/patient.js [hunk 5]
```diff
@@ -192,7 +198,13 @@ function handleFetchPatient(uuid) {
     'ON p.debitor_uuid = d.uuid AND d.group_uuid = dg.uuid AND p.project_id = proj.id ' +
     'WHERE p.uuid = ?';
 
-  return db.exec(patientDetailQuery, [uuid]);
+  return db.exec(patientDetailQuery, uuid)
```
Comment (jniles) [inline server/controllers/medical/patient.js:201]: Why did you need to change code in the patient controller for a cashbox pull request?

File: server/controllers/finance/cashboxes.js [hunk 6]
```diff
@@ -32,6 +32,16 @@ exports.list = function list(req, res, next) {
   sql =
     'SELECT id, text FROM cash_box ';
 
+  if(req.query.full === '1') {
```
Comment (jniles) [inline server/controllers/finance/cashboxes.js:35]: We should talk about this design Monday.  I think I know a better way.

File: server/controllers/finance/cashboxes.js [hunk 7]
```diff
@@ -52,6 +62,8 @@ exports.list = function list(req, res, next) {
 
   sql += ';';
 
+  console.log('sql', sql);
```
Comment (jniles) [inline server/controllers/finance/cashboxes.js]: Please remove `console.log()`s before submitting a PR.

File: client/src/i18n/fr.json [hunk 8]
```diff
@@ -317,6 +317,11 @@
    "WARNING"          : "Ce patient est un conventionné et ne peut payer à la caisse auxilliare",
    "WARNING_2"        : "Pour le paiement des anciennes factures il faudrait changer son groupe débiteur "
    },
+"CASH_BOX_TYPE" : {
```
Comment (jniles) [inline client/src/i18n/fr.json:320]: Don't these keys already exist somewhere in the fr.json file?

**Comment by jniles** [comment]
I've completed an initial review.  Here are a couple general comments:
1. There is no need to implement `is_bank` - we shouldn't use cashboxes to module banks.  If we need a separate `bank` entity, we can make it.  For now, we should be able to do everything with journal vouchers.
2. For query strings, it is much easier to use Angular's automatic query string templating.  By default, it uses [$httpParamSeralizer](https://docs.angularjs.org/api/ng/service/$httpParamSerializer) to perform templating.  An example usage looks like this:
   
   ``` js
   // vm.model might map to an <input type="checkbox" ng-model="$ctrl.model.is_auxillary" ng-true-value="1" ng-false-value="0">
   vm.model = { is_auxillary : 0 };
   
   function submit(invalid) {
     if (invalid) { return; }
   
    // this should be done using a service, but I am using $http as an example
     $http.get('/cashboxes', { params : vm.model })
     .then(util.unwrapHttpResponse)
     .then(function (cashboxes) {
       vm.cashboxes = cashboxes;
     });
     // remember to handle errors!
   }
   
   // .... on the server ...
   cashboxes.list = function list(req, res, next) {
     if (req.query.is_auxillary) {
      // logic!
     }
   }
   ```
3. Instead of using `$ctrl.search()` to filter values on the client, you could have `$ctrl.search()` look up filtered values from the server.  It isn't very necessary for the cashboxes component (we expect to have few cashboxes), but it will allow all our `find-X` components to look similar and be easier to debug + maintain + test.  I think you will also be able to reduce your controller code by about half with this change.  [Here is the documentation for typeaheads](https://angular-ui.github.io/bootstrap/#/typeahead).
4.  We should talk about pattern of looking up cashboxes + currencies.  I don't think it is necessary to use the `full=1` route just yet, but we can discuss it on Monday.

File: client/src/partials/templates/bhCashBoxFinder.tmpl.html (hunk 4 above)
Comment (DedrickEnc) [inline client/src/partials/templates/bhCashBoxFinder.tmpl.html:23]: I don't need to send a`http` request every time a user will press on a keyboard, since I am dealing with the cash box, a regex is optimal.

File: server/controllers/medical/patient.js (hunk 5 above)
Comment (DedrickEnc) [inline server/controllers/medical/patient.js:201]: I began first of all to try adding new interfaces, but I notices that every needed interfaces were present, so I continued directly implementing the cash box component.
This change was made at the very beginning.

File: server/controllers/finance/cashboxes.js (hunk 6 above)
Comment (DedrickEnc) [inline server/controllers/finance/cashboxes.js:35]: excellent!

File: client/src/i18n/fr.json (hunk 8 above)
Comment (DedrickEnc) [inline client/src/i18n/fr.json:320]: No, according to my knowledge.

Comment (jniles) [inline client/src/i18n/fr.json:320]: :+1:

File: server/controllers/medical/patient.js (hunk 5 above)
Comment (jniles) [inline server/controllers/medical/patient.js:201]: So...  are these changes necessary?  What did you change and why?

**Comment by DedrickEnc** [comment]
@jniles

**Comment by jniles** [comment]
LGTM.  As soon as this is rebased and all the tests pass it can be merged.
